### PR TITLE
[RFC] Object Field

### PR DESCRIPTION
**Pull request by drewgatchell**
New RFC to add an Top-Level `object` field to support Cloud SaaS platform use cases.

### CODE REVIEW

**Comment by ebeahan** [comment]
Thanks for the RFC submission @drewgatchell! The RFC process for ECS is still new, so please feel free to share any feedback on the RFC docs to benefit future contributors 😄 .  

Can you elaborate a bit on how `object` is focused on SaaS providers? I agree many SaaS providers do expose an API interface (typically HTTP) for CRUD-type operations. However, so do many other types of systems/services and for me it wasn't immediately clear the relationship being established in the proposal. Do you have any basic examples you could share that might provide a bit more detail?

Using `object` as a field set name could be overloading the term a bit. `object` already exists as a [data type](https://www.elastic.co/guide/en/elasticsearch/reference/current/object.html) and all the ECS field sets are [objects themselves](https://www.elastic.co/guide/en/ecs/current/ecs-faq.html#dot-notation). Naming a top-level ECS field `object` could create some user confusion.

**Comment by drewgatchell** [comment]
Hi @ebeahan; totally understand on the concern of overloading the term, I'm certainly open to suggestions on better terminology.

I don't mean to imply this can/should only be used for SaaS applications, it's just in my current use-case around SaaS eventing is the driver for the RFC. That said, below are some examples that can hopefully provide some clarity.

First Up is a Zoom "Meeting Created" event:

`{
  "event": "meeting.created",
  "payload": {
    "account_id": "o8KK_AAACq6BBEyA70CA",
    "operator": "someemail@email.com",
    "operator_id": "uLoRgfbbTayCX6r2Q_qQsQ",
    "object": {
      "uuid": "czLF6FFFoQOKgAB99DlDb9g==",
      "id": 111111111,
      "host_id": "uLoRgfbbTayCX6r2Q_qQsQ",
      "topic": "My Meeting",
      "type": 2,
      "start_time": "2019-07-09T17:00:00Z",
      "duration": 60,
      "timezone": "America/Los_Angeles"
    }
  }
}`

Next Up a Salesforce Lightning URI Event (essentially a "Read" operation on a record):
`{"attributes":{"type":"LightningUriEvent","url":"/services/data/v48.0/sobjects/LightningUriEvent/000000000000000AAA"},"AppName":"one:one","CreatedDate":"2020-07-09T20:19:09.542+0000","DevicePlatform":"SFX:BROWSER:DESKTOP","DeviceSessionId":"4975ffc1f7da123028b197405974763f8bd64105","Duration":1909.0,"EffectivePageTime":1909.0,"EventDate":"2020-07-09T20:19:04.905+0000","EventIdentifier":"000a4ec7-c502-4c0b-9839-725f72904c8f","LoginKey":"7JIrlZKYOvGmEW7L","Operation":"Read","OsName":"WINDOWS","OsVersion":"10","PageStartTime":"2020-07-09T20:15:24.712+0000","PageUrl":"/lightning/r/custom_page__c/a5I3q000000AbOSRD0/view","PreviousPageAppName":"Custom_App","PreviousPageEntityId":"a0i3q00000vGJKLAA4","PreviousPageEntityType":"other_page__c","PreviousPageUrl":"/lightning/r/other_page__c/a0i3q00000vGJKLAA4/view","QueriedEntities":"custom_page__c","RecordId":"a5I3q000000AbOSRD0","SessionKey":"Ew6Tjg1vwtFsNZ/y","SessionLevel":"STANDARD","SourceIp":"255.255.255.255","UserId":"0053q00000ABUPIAA5","Username":"test_user@example.com","UserType":"Standard"}`

A Zendesk Audit Log Example:
`{
  "id":           498483,
  "actor_id":     1234,
  "source_id":    3456,
  "source_type":  "user",
  "source_label": "John Doe",
  "action":       "update",
  "action_label": "Updated"
  "changes_description": "Role changed from Administrator to End User"
  "ip_address":   "209.119.38.228",
  "created_at":  "2012-03-05T11:32:44Z",
  "url":          "https://company.zendesk.com/api/v2/audit_logs/498483.json",
}`

So for all these events; some user is taking a CRUD action on an object; Zoom they're creating a meeting, Salesforce they're reading a Record; and with Zendesk they're updating a user role. For the most part there's no real good place to drop most of this information in ECS other than to just drop it in some top-level custom object (similar to what's being done for the existing O365 Filebeat module). Which isn't necessarily horrible approach, but what if my company uses both Zoom and Google Meet (as many do) and we're collecting logs for both and we want to track user activity across both? Well now I need to know the field names for both platforms and adjust my query as such; or I can use a normalized `object` field and have my query filter be something like `object.type = meeting'. On a similar note, what if my company leverages Salesforce, ServiceNow, and Workday and I want to attempt to detect insider threat activity that crosses all three platforms, it's much simpler to pivot off a normalized field that highlights "records". This issue only gets more prominent as you encounter the fact that some of these platforms have multiple schemas for simply documenting the same type of CRUD activity; so now it's not just a matter of adjusting your search to account for just 3 platforms, but 3 + n variations.

I hope this brings some clarity to what I'm trying to accomplish.

**Comment by ebeahan** [comment]
Thanks for the additional details and examples!

Many of the points you touched on are indeed what ECS aims to help solve: a consistent, standardize way to structure data to facilitate analyzing sources from different, diverse data sources. I'm still a bit fuzzy on what challenges the proposed `object` field set would be solving for; many of the challenges you mentioned could be solved with existing field sets.

> but what if my company uses both Zoom and Google Meet (as many do) and we're collecting logs for both and we want to track user activity across both?

Have you explored adopting the existing [`user.*`](https://www.elastic.co/guide/en/ecs/current/ecs-user.html) and [`related.user`](https://www.elastic.co/guide/en/ecs/current/ecs-related.html) fields? Normalizing to common `user` fields (top-level or nested) is intended to address this type of scenario.

> I can use a normalized object field and have my query filter be something like `object.type = meeting'.

The [categorization fields](https://www.elastic.co/guide/en/ecs/current/ecs-category-field-values-reference.html) are intended to deterministically help classify events (although `meeting` isn't currently an allowed category). Adding [`labels`](https://www.elastic.co/guide/en/ecs/master/ecs-base.html) aligning with a predetermined convention could be another approach.

Looking at one of the examples:

```json
{ "id": 498483, "actor_id": 1234, "source_id": 3456, "source_type": "user", "source_label": "John Doe", "action": "update", "action_label": "Updated" "changes_description": "Role changed from Administrator to End User" "ip_address": "209.119.38.228", "created_at": "2012-03-05T11:32:44Z", "url": "https://company.zendesk.com/api/v2/audit_logs/498483.json", }
```

I see opportunity to align with existing ECS fields such as `@timestamp`, `source.ip`, `url.original`, `event.category: iam`, etc. What do envision the `object` field set solving for with this type of event beyond ECS's current spec?

**Comment by drewgatchell** [comment]
Hi @ebeahan;

So this isn't aimed at figuring out how to normalize data present in those events where there are existing ECS fields; it's great that I can normalize `user` into the existing `user.*` fields. I can use that to pivot off a specific user, but what if I want to know all the participants of a specific meeting? I'd have to pivot off a custom field, or what if I want to look at multiple meetings that span multiple services, and I want my query to be more targeted than "give me all events labeled as 'meeting'"? Now my pivot looks at multiple custom fields (at least one for each service I'm targeting if I'm pivoting off a known set of "meeting ids").

In the example you highlighted in your post; I can normalize `actor_id` to the `user.id` field; in addition to the other fields you mentioned, but if I wanted to target the object that was changed (`source_id`); I'd have to dig down into a custom field; because there's no logical place that I currently know of in ECS to document that. 

A similar circumstance in the SaaS world would be a file modification/update on something residing in Box/Dropbox/GDrive/etc. If a file is modified, I can drop the file details in the `file.*` fields. There is no similar field for anything else that's not a file on a SaaS platform within ECS today; that's essentially the problem the `object.*` field is looking to solve; the ability to pivot on a normalized field so that I don't need to know the specific implementations of eventing for every SaaS provider when I want to pivot on the "object" (or "subject" if you will) of the event.

As noted in my previous post, this problem isn't necessarily isolated to trying to pivot across multiple SaaS platforms. Some platforms (Salesforce being one, Box being another) have multiple schemas for events, so the "Object" of the event might be in one of 5 different fields depending on which schema the event you received is using. Slightly annoying to have to write a query that goes `object_id = 12345 or objectid = 12345 or recordid = 12345` (repeat ad nauseam); so I'm hoping implementing this will make it so that type of query clause becomes `object.id=12345`.

File: rfcs/0000-object-field.md [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+-->
+This RFC calls for the introduction of a top-level `object` field with an initial fieldset of the following child fields:
+
+| field | type | description |
+| --- | --- | --- |
+| `object.name` | keyword | Name of the Object |
+| `object.id` | keyword | ID of the Object |
+| `object.type` | keyword | Type of Object Represented (e.g. `record`, `meeting`, `repository`, `organization`, etc.) |
+| `object.additional_details` | object | Custom Key/Value pairs representing data relevant to the object |
+| `object.owner.id` | keyword | ID of the Object Owner |
```
Comment (bkimmel) [inline rfcs/0000-object-field.md:25]: :grey_question: How do these `owner` fields overlap/co-exist with https://www.elastic.co/guide/en/ecs/current/ecs-user.html ?

Comment (drewgatchell) [reply rfcs/0000-object-field.md:25]: It's very likely that owner could just be a nested user object; I haven't encountered an instance yet where that wouldn't work.

**Comment by drewgatchell** [comment]
This isn't reflected in the RFC as currently written, but I just recently encountered this situation; so I want to open it up for comments before adjusting the RFC. Some platforms can push an event where an action was taken on multiple "objects" (e.g. GitHub pushes an event when an App Install is removed from multiple repos or added to multiple repos). So perhaps the "object" field (or whatever it will end up being named, given the concern about overloading) can be an array of "objects". I know there are some fields in ECS that are arrays of keywords (tag, the related.* fields, etc); but I can't recall if there are any fields that are arrays of objects.

**Comment by webmat** [comment]
Forgot to respond to your latest comment on arrays of objects.

Arrays of objects are supported in Elasticsearch, but they come with caveats that are discussed at the top of [the "nested" documentation](https://www.elastic.co/guide/en/elasticsearch/reference/current/nested.html). This data type would likely be better to use for this situation.

However I'd like to proceed cautiously here. While Elasticsearch will handle `nested` beautifully, Kibana will still struggle displaying those fields, or their subfields.

We do want to investigate this in full and come up with a recommended approach that's applicable for the many situations where we need arrays of objects (#857). In addition to guidance, this work may even end up driving changes in the Elastic Stack to improve support across the board. We'll see.

But for now I think we should focus on the single "object" case, which is simpler and will be useful for most situations, IMO.

File: rfcs/0000-object-field.md [hunk 2]
```diff
@@ -0,0 +1,118 @@
+# 0000: Object Fields
```
[the remark this answers is not in the document]
Comment (drewgatchell) [reply rfcs/0000-object-field.md]: +1 for resource or entity

[the remark this answers is not in the document]
Comment (webmat) [reply rfcs/0000-object-field.md]: Just came across another good one:

* asset

[the remark this answers is not in the document]
Comment (drewgatchell) [reply rfcs/0000-object-field.md]: asset isn't too bad; I'm not sure I like it as much as resource or entity, but that's mostly due to my first thought when I see/hear "asset" is something more physical/tangible like a laptop/desktop/server.

[the remark this answers is not in the document]
Comment (ebeahan) [reply rfcs/0000-object-field.md]: Zooming out another level, perhaps we could explore `response` which could contain multiple `resources` or `entities` (idea came from looking at how the OpenAPI spec describes an API [response operation](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.3.md#responseObject)).

File: rfcs/0000-object-field.md (hunk 1 above)
Comment (bkimmel) [reply rfcs/0000-object-field.md]: > In general user.* at the root of the event is the "doer". If we need to capture an object's owner, we could nest the user fields in an additional location. 

@webmat That makes a lot of sense, but it's hard to discern the "doer" inclination from the description of `user.*` which reads "The user fields describe information about the user that is relevant to the event." The spec might benefit from making that distinction more explicit to the reader.

Comment (webmat) [reply rfcs/0000-object-field.md]: Agreed. We'll be improving on this soon.

**Comment by drewgatchell** [comment]
> Once again, thanks @drewgatchell for submitting this.
> 
> We're on the way to identifying a sponsor for this at Elastic. This has already sparked interest internally.
> 
> If you're willing to keep working on this, you're welcome to open the stage 1 pull request. If not, your feedback will of course be welcome on the following stage PRs :-)
> 
> In stage 1, we'll want to address some of the following:
> 
> * Converge on a name and adjust the RFC accordingly. So far "resource" makes most sense, I think.  But still open for discussion.
> * Include real world documents in the RFC document itself.
> * Caveats: Mention the names that were considered.
> * Cloudiness: this doesn't relate only to cloud systems. I think this will be applicable to on prem as well. As such, we could mention this may be applicable to Active Directory events, and review the text to tone down the relation to cloud/SaaS.
> * Start discussing how to handle "resource" vs other supported resources in ECS (e.g. "user"): should we fill both or is this mutually exclusive?
> * Officially identify the sponsor (that one's on us :-) )

Just want to follow up on this; I will definitely work on the Stage 1 request; I'm excited to see this move forward.

**Pull request merged**